### PR TITLE
Modify text visualizer to better handle newlines and unprintable control characters

### PR DESCRIPTION
**Pull request by PathogenDavid**
This change replaces `\n` (and other more esoteric Unicode newlines) with a space.

Carriage returns are dropped entirely (with the assumption they'll come with line feeds and that for the purposes of this visualization nobody will care about a bare `\r`.)

Other control characters (except `\t`) are replaced with � (the Unicode replacement character.)

Here's what that looks like with some test strings:

image

Strings are now processed in batches when possible to take advantage of reusing the temporary `StringBuilder`. This also allows us to skip processing of values which will never be displayed in the case of high-frequency inputs.

The design of this PR is such that future specializations of this visualizer (as mentioned in https://github.com/bonsai-rx/bonsai/issues/1670#issuecomment-2083904942) can simply override `MakeDisplayString` to change how the display string processing is done.

I also added disposing of `textPanel`, which was previously missing.

As an aside, I still implemented `Show(object)` since it's publicly accessible, but I did notice that the visualizer's actual text might not actually be updated in situations where someone was calling it directly and no batches come in to `ShowBuffer`. (This edge case wasn't handled before either.) Maybe the implementation should throw or assert to discourage using it? (Or if there's practical reasons to use it, maybe we should have it update the display anyway even though that'd be very inefficient.)

Fixes https://github.com/bonsai-rx/bonsai/issues/1670

### CODE REVIEW

File: Bonsai.Design/ObjectTextVisualizer.cs [hunk 1]
```diff
@@ -30,27 +31,90 @@ public class ObjectTextVisualizer : BufferedVisualizer
...
+            var rawText = value?.ToString() ?? string.Empty;
+            var stringBuilder = new StringBuilder(rawText.Length);
+            MakeDisplayString(rawText, stringBuilder);
+            DisplayString(stringBuilder.ToString());
+        }
+
+        /// <summary>Converts the specified string for display in this visualizer</summary>
+        /// <param name="rawText">The raw text to be converted</param>
+        /// <param name="stringBuilder">The string builder which receives the display string</param>
+        protected virtual void MakeDisplayString(string rawText, StringBuilder stringBuilder)
```
Comment (glopesdev) [inline Bonsai.Design/ObjectTextVisualizer.cs:78]:
```suggestion
        /// <summary>
        /// Appends the display text for the specified object to the text buffer.
        /// </summary>
        /// <param name="stringBuilder">The string builder which receives the display text.</param>
        /// <param name="value">The object for which to retrieve the display text.</param>
        protected virtual void AppendDisplayText(StringBuilder stringBuilder, object value)
```

Since we are making this part of the API, I was considering we could refactor the method name to the above for consistency with `StringBuilder` methods and also for similarity with [`GetDisplayText`](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.design.collectioneditor.getdisplaytext) which we use elsewhere for similar purposes.

The idea behind passing here the actual `object` is to give more flexibility for future specialized text visualizers which might want to access specific instance fields or properties of the object directly, e.g. reflection, etc for debugging purposes. Looking at the code I was hoping it shouldn't impact performance too much since `StringBuilder` should resize efficiently in most cases.

File: Bonsai.Design/ObjectTextVisualizer.cs [hunk 2]
```diff
@@ -30,27 +31,90 @@ public class ObjectTextVisualizer : BufferedVisualizer
...
+                        stringBuilder.Append('\xFFFD');
+                        break;
+                    default:
+                        stringBuilder.Append(c);
+                        break;
+                }
             }
         }
 
+        private void DisplayString(string value)
```
Comment (glopesdev) [inline Bonsai.Design/ObjectTextVisualizer.cs:110]:
```suggestion
        private void AppendString(string value)
```

This is just a nitpick since it is a private method, but when reading the code I was being thrown off by the name since the method doesn't actually display the string.

File: Bonsai.Design/ObjectTextVisualizer.cs [hunk 3]
```diff
@@ -30,27 +31,90 @@ public class ObjectTextVisualizer : BufferedVisualizer
...
         {
-            value ??= string.Empty;
-            var text = value.ToString();
-            text = Regex.Replace(text, @"\r|\n", string.Empty);
-            buffer.Enqueue(text);
-            while (buffer.Count > bufferSize)
+            var rawText = value?.ToString() ?? string.Empty;
+            var stringBuilder = new StringBuilder(rawText.Length);
+            MakeDisplayString(rawText, stringBuilder);
+            DisplayString(stringBuilder.ToString());
```
Comment (glopesdev) [inline Bonsai.Design/ObjectTextVisualizer.cs:72]: Do we need to update the text box and invalidate the text panel here?

Although I agree we probably should just throw here. This code should never be called anyway now that `ShowBuffer` takes care of everything. The idea for buffered visualizers is by default they just call `Show` successively for each value, but since we have a more optimized implementation here this is not used.

Derived classes can just override it again if they need to do something with it.

File: Bonsai.Design/ObjectTextVisualizer.cs [hunk 4]
```diff
@@ -30,27 +31,90 @@ public class ObjectTextVisualizer : BufferedVisualizer
...
+            for (int i = Math.Max(0, values.Count - bufferSize); i < values.Count; i++)
             {
-                base.ShowBuffer(values);
-                textBox.Text = string.Join(Environment.NewLine, buffer);
-                textPanel.Invalidate();
+                var rawText = values[i].Value?.ToString() ?? string.Empty;
+                sb.Clear();
+                sb.EnsureCapacity(rawText.Length);
+                MakeDisplayString(rawText, sb);
+                DisplayString(sb.ToString());
```
Comment (glopesdev) [inline Bonsai.Design/ObjectTextVisualizer.cs:50]:
```suggestion
                sb.Clear();
                AppendDisplayText(sb, values[i].Value);
                AppendString(sb.ToString());
```

Given all the above, I guess this bit would look like this. I was wondering here whether we will ever want to provide the timestamp, but I guess if we ever need that we can later make a non-breaking change by introducing a new virtual method `AppendTimestampedDisplayText` which receives the timestamped value and by default just calls `AppendDisplayText` with the value.

**Review by glopesdev — CHANGES_REQUESTED** [review]
Looks great, just added a few refactoring suggestions for future-proofing more complex text visualizers and naming consistency.

File: Bonsai.Design/ObjectTextVisualizer.cs (hunk 1 above)
Comment (PathogenDavid) [reply Bonsai.Design/ObjectTextVisualizer.cs:78]: I wasn't super happy with the name so this is much better, thanks!

I actually just moved the `EnsureCapacity` call to inside `AppendDisplayText` since that's really where it belonged anyway, so it really only leads to an extra allocation in the `Show(object)` case which we don't really care about.

File: Bonsai.Design/ObjectTextVisualizer.cs (hunk 4 above)
Comment (PathogenDavid) [reply Bonsai.Design/ObjectTextVisualizer.cs:50]: Yeah, providing the timestamp to `AppendDisplayText` just opens another weird thing to handle in the `Show(object)` edge case.

I think if we're making a text visualizer that also shows timestamps then at that point we need to revisit making this visualizer have configuration.

File: Bonsai.Design/ObjectTextVisualizer.cs [hunk 5]
```diff
@@ -30,25 +31,99 @@ public class ObjectTextVisualizer : BufferedVisualizer
...
+                buffer.Dequeue();
+
+            // Add new values to the buffer (and only the ones which might appear)
+            for (int i = Math.Max(0, values.Count - bufferSize); i < values.Count; i++)
             {
-                base.ShowBuffer(values);
-                textBox.Text = string.Join(Environment.NewLine, buffer);
-                textPanel.Invalidate();
+                sb.Clear();
+                AppendDisplayText(sb, values[i]);
```
Comment (glopesdev) [inline Bonsai.Design/ObjectTextVisualizer.cs:47]:
```suggestion
                AppendDisplayText(sb, values[i].Value);
```

To keep the current default value we need to get the value here to avoid printing the timestamp.

File: Bonsai.Design/ObjectTextVisualizer.cs [hunk 6]
```diff
@@ -30,25 +31,99 @@ public class ObjectTextVisualizer : BufferedVisualizer
...
+            stringBuilder.EnsureCapacity(stringBuilder.Length + rawText.Length);
+
+            foreach (var c in rawText)
             {
+                switch (c)
+                {
+                    // Carriage returns are presumed to be followed by line feeds, skip them entirely
+                    case '\r':
+                        continue;
+                    // Newlines become space so that things like multi-line JSON or matricies are still visible
```
Comment (glopesdev) [inline Bonsai.Design/ObjectTextVisualizer.cs:84]:
```suggestion
                    // Newlines become space so that things like multi-line JSON or matrices are still visible
```

File: Bonsai.Design/ObjectTextVisualizer.cs [hunk 7]
```diff
@@ -30,25 +31,99 @@ public class ObjectTextVisualizer : BufferedVisualizer
...
+                    case '\n':
+                    case '\x0085': // Next line character (NEL)
+                    case '\x2028': // Unicode line separator
+                    case '\x2029': // Unicode paragraph separator
+                        stringBuilder.Append(' ');
+                        break;
+                    case '\t':
+                        stringBuilder.Append(c);
+                        break;
```
Comment (glopesdev) [inline Bonsai.Design/ObjectTextVisualizer.cs:93]: Can we move this case next to `default`, or will it break the switch logic? Although I guess we could probably also leave it as it is for readability.

**Review by glopesdev — CHANGES_REQUESTED** [review]
Just a few final corrections.

File: Bonsai.Design/ObjectTextVisualizer.cs (hunk 7 above)
Comment (PathogenDavid) [reply Bonsai.Design/ObjectTextVisualizer.cs:93]: It needs to come before the control characters block since tabs are control characters. I think it's better to show the edge case is explicitly handled.

File: Bonsai.Design/ObjectTextVisualizer.cs (hunk 5 above)
Comment (PathogenDavid) [reply Bonsai.Design/ObjectTextVisualizer.cs:47]: Oops, that's what I get for changing my test workflow. (I added a timer and thought the timestamps were what it was emitting, didn't notice they were on the other lines too.)

**Comment by PathogenDavid** [comment]
Final corrections done. Let me know if you'd prefer I squash the two PR address commits together. (I normally would since them being separate isn't helpful, but I know it drives some people up the wall to have force pushes in a reviewed PR.)

**Comment by glopesdev** [comment]
@PathogenDavid squashing and force pushing is fine 👍

**Comment by PathogenDavid** [comment]
Done! (Ended up squishing it all down into a single commit since both addressing commits were pretty inconsequential.)

**Review by glopesdev — APPROVED** [review]
